tmc-plugin/src/fi/helsinki/cs/tmc/model/CourseDb.java

```diff
@@ -85,25 +87,53 @@ public void setCurrentCourseName(String currentCourseName) {
         save();
     }
 
-    public List<Theme> getCurrentCourseThemes() {
-        return getCurrentCourse().getThemes();
+    public int getCurrentCourseWeekCount() {
+        return getCurrentCourse().getWeekCount();
+    }
+
+    public List<Exercise> getExercisesByWeek(int week) {
+        return getCurrentCourse().getExercisesByWeek(week);
+    }
+
+    public int getPresumedCurrentWeek() {
+        Project project = OpenProjects.getDefault().getOpenProjects()[0];
+        ProjectMediator projectMediator = ProjectMediator.getInstance();
+        Exercise exercise = projectMediator
+            .tryGetExerciseForProject(projectMediator.wrapProject(project),
+                CourseDb.getInstance());
+        return exercise.getWeek();
     }
 
-    public Theme themeForExercise(Exercise exercise) {
-        List<Theme> currentCourseThemes = getCurrentCourseThemes();
-        if (currentCourseThemes == null) {
-            return new Theme("no theme");
+    public int getCurrentWeek() {
+        for (int i = getCurrentCourseWeekCount(); i > 0; i--) {
+            for (Exercise ex : getExercisesByWeek(i)) {
+                if (ex.isAttempted()) {
+                    return i;
+                }
+            }
         }
-        for (Theme theme : currentCourseThemes) {
-            if (theme.shouldContain(exercise)) {
-                return theme;
+        return 1; //default
+    }
+
+    public int getAdaptiveWeek() {
+        int week = -1;
+        for (int i = 1; i <= getCurrentCourseWeekCount(); i++) {
+            boolean wholeWeekDone = true;
+            for (Exercise ex : getExercisesByWeek(i)) {
+                if (!ex.isCompleted()) {
+                    wholeWeekDone = false;
+                    break;
+                }
+            }
+            if (wholeWeekDone) {
+                week = i;
```

Comment:
Adaptive exercises are available only after a fully completed week. This method shouldn't mind if a user left one week undone. We haven't really had the opportunity to test it though... :(